bftengine/src/bftengine/ReplicaFactory.cpp

```diff
@@ -0,0 +1,290 @@
+        replicaConfig.numReplicas,
+        replicaConfig.fVal,
+        replicaConfig.cVal,
+        replicaConfig.numReplicas + replicaConfig.numRoReplicas + replicaConfig.numOfClientProxies +
+            replicaConfig.numOfExternalClients + replicaConfig.numOfClientServices + replicaConfig.numReplicas,
+        replicaConfig.clientBatchingMaxMsgsNbr));
+    unique_ptr<MetadataStorage> metadataStoragePtr(metadataStorage);
+    auto objectDescriptors = static_cast<PersistentStorageImp *>(persistentStoragePtr.get())
+                                 ->getDefaultMetadataObjectDescriptors(numOfObjects);
+    isNewStorage = metadataStoragePtr->initMaxSizeOfObjects(objectDescriptors, numOfObjects);
+    // Check if we need to remove the metadata or start a new epoch
+    bool erasedMetaData = false;
+    uint32_t actualObjectSize = 0;
+    metadataStoragePtr->read(ConstMetadataParameterIds::ERASE_METADATA_ON_STARTUP,
+                             sizeof(erasedMetaData),
+                             (char *)&erasedMetaData,
+                             actualObjectSize);
+    bool startNewEpoch = false;
+    metadataStoragePtr->read(
+        ConstMetadataParameterIds::START_NEW_EPOCH, sizeof(startNewEpoch), (char *)&startNewEpoch, actualObjectSize);
+    if (startNewEpoch) {
+      bftEngine::EpochManager::instance().startNewEpoch();
+      LOG_INFO(GL, "We should start a new epoch");
+    }
+    LOG_INFO(GL, KVLOG(erasedMetaData));
+    if (erasedMetaData) {
+      // Here when metadata is erased, we need to update DBCheckpointManager.
+      DbCheckpointManager::instance().setIsMetadataErased(true);
+      metadataStoragePtr->eraseData();
+      isNewStorage = metadataStoragePtr->initMaxSizeOfObjects(objectDescriptors, numOfObjects);
+      auto secFileDir = ReplicaConfig::instance().getkeyViewFilePath();
+      LOG_INFO(GL, "removing " << secFileDir << " files if exist");
+      try {
+        for (auto &it : fs::directory_iterator{secFileDir}) {
+          if (it.path().string().find(secFilePrefix) != std::string::npos) {
+            fs::remove(it.path());
+          }
+        }
+      } catch (std::exception &e) {
+        LOG_FATAL(GL, "unable to remove the secret file, as we erased the metadata we won't be able to restart");
+      }
+    }
+
+    // Init the persistent storage
+    ((PersistentStorageImp *)persistentStoragePtr.get())->init(std::move(metadataStoragePtr));
+  }
+  auto replicaInternal = std::make_unique<ReplicaInternal>();
+  shared_ptr<MsgHandlersRegistrator> msgHandlersPtr(new MsgHandlersRegistrator());
+  auto incomingMsgsStorageImpPtr =
+      std::make_unique<IncomingMsgsStorageImp>(msgHandlersPtr, timersResolution, replicaConfig.replicaId);
+  auto &timers = incomingMsgsStorageImpPtr->timers();
+  shared_ptr<IncomingMsgsStorage> incomingMsgsStoragePtr{std::move(incomingMsgsStorageImpPtr)};
+  shared_ptr<bft::communication::IReceiver> msgReceiverPtr(new MsgReceiver(incomingMsgsStoragePtr));
```

Comment:
`std::make_shared<bft::communication::IReceiver>(...)` is preferable. Few reasons why.